source/s3_auto_ranged_put.c

```diff
@@ -800,23 +866,71 @@ static int s_s3_auto_ranged_put_prepare_request(
 
             if (request->num_times_prepared == 0) {
-                if (s_skip_parts_from_stream(
-                        meta_request,
-                        auto_ranged_put->prepare_data.num_parts_read_from_stream,
-                        request->part_number - 1)) {
+                if (auto_ranged_put->has_content_length && s_skip_parts_from_stream(
+                                                               meta_request,
+                                                               auto_ranged_put->prepare_data.num_parts_read_from_stream,
+                                                               request->part_number - 1)) {
                     goto message_create_failed;
                 }
-                auto_ranged_put->prepare_data.num_parts_read_from_stream = request->part_number - 1;
 
-                aws_byte_buf_init(&request->request_body, meta_request->allocator, request_body_size);
+                if (aws_s3_meta_request_body_has_no_more_data(meta_request)) {
+                    request->is_noop = true;
 
-                if (aws_s3_meta_request_read_body(meta_request, &request->request_body)) {
-                    goto message_create_failed;
+                    AWS_LOGF_DEBUG(
+                        AWS_LS_S3_META_REQUEST,
+                        "id=%p UploadPart with part num %u for Multi-part Upload, with ID:%s"
+                        "is noop due to encountering end of stream",
+                        (void *)meta_request,
+                        request->part_number,
+                        aws_string_c_str(auto_ranged_put->upload_id));
+
+                } else {
+                    auto_ranged_put->prepare_data.num_parts_read_from_stream = request->part_number - 1;
+
+                    aws_byte_buf_init(&request->request_body, meta_request->allocator, request_body_size);
+
+                    if (aws_s3_meta_request_read_body(meta_request, &request->request_body)) {
+                        goto message_create_failed;
+                    }
+                    ++auto_ranged_put->prepare_data.num_parts_read_from_stream;
+
+                    /* BEGIN CRITICAL SECTION */
+                    {
+                        aws_s3_meta_request_lock_synced_data(meta_request);
+
+                        ++auto_ranged_put->synced_data.num_parts_prepared;
+
+                        auto_ranged_put->prepare_data.is_body_stream_at_end =
+                            aws_s3_meta_request_body_has_no_more_data(meta_request);
+
+                        struct aws_byte_buf checksum_buf = {0};
+                        aws_array_list_set_at(
+                            &auto_ranged_put->encoded_checksum_list, &checksum_buf, request->part_number - 1);
+
+                        struct aws_string *null_etag = NULL;
+                        aws_array_list_set_at(
+                            &auto_ranged_put->synced_data.etag_list, &null_etag, request->part_number - 1);
```

Comment:
added comment. other code sets the etag/checksum based on the data returned from s3 for pause/resume flows, and there we can have gaps in completed parts. here its just zeroing out values for the new parts as they are prepared to avoid any garbage values during error flows